### PR TITLE
[Data objects] Add default value generator service

### PR DESCRIPTION
**Pull request by BlackbitNeueMedien**
Currently some fields support default values, others don't - but in any case these default values are static. This PR adds an input field in field definition to set a default value generator class or service. Use-cases are e.g. URL generation based on other fields of the object, Article number generation etc. - things which most often get only generated once per object and then stay the same forever.

The specified default value generator generates a value on saving if the field is empty.

This PR only implements this for input fields but it is easily extendable to all field types (except those where it does not make sense - calculated value e.g.).

PS: As soon as #5027 gets merged I will refactor the `DefaultValueGeneratorInterface` to also use `ClassResolver`.

### CODE REVIEW

File: models/DataObject/ClassDefinition/Helper/DefaultValueGeneratorResolver.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace Pimcore\Model\DataObject\ClassDefinition\Helper;
+
+use Pimcore\Logger;
+use Pimcore\Model\DataObject\ClassDefinition\DefaultValueGeneratorInterface;
+
+class DefaultValueGeneratorResolver
```
Comment (dpfaffenbauer) [inline models/DataObject/ClassDefinition/Helper/DefaultValueGeneratorResolver.php]: why not use `ClassResolver`?

Comment (BlackbitNeueMedien) [reply models/DataObject/ClassDefinition/Helper/DefaultValueGeneratorResolver.php]: This is part of #5027 which hat not been merged yet.

Comment (dpfaffenbauer) [reply models/DataObject/ClassDefinition/Helper/DefaultValueGeneratorResolver.php]: true, so we should wait for #5027 to be merged before this

**Comment by BlackbitNeueMedien** [comment]
I added context information to the Field definition class similar to `CalculatedValue`'s `ownerType`, `ownerName`, `index` etc. as these information can be quite useful for generating values.
Sadly I fail to get the name of the brick container field (in `CalculatedValue` this would be `ownerName`, see https://pimcore.com/docs/6.x/Development_Documentation/Objects/Object_Classes/Data_Types/Calculated_Value_Type.html#page_Context-Information-for-Calculation-Class). Any ideas how to get that?

**Comment by weisswurstkanone** [comment]
I guess somehow related to
https://github.com/pimcore/pimcore/issues/4600

Can you give a judgement on what the consequences would be? If I interpret your PR correctly then it wouldn't be possible to null it out anymore?

**Comment by BlackbitNeueMedien** [comment]
Now the default value is generated on loading and saving. That means if the object has an empty input field, it is tried to generate a value and display it in the editing panel  (`getDataForEditmode`). This is similar to what current static default values do.

If the user deletes the content, the default value gets generated (`getDataForResource`). This is the thing which current static default values do not so (and is addressed in #4600)

In my opinion it is ok to not be able to set the field empty, isn't it? 
If we remove the generation when the object is getting saved (`getDataForResource`), then it is possible to nullify the field but the behaviour is somehow strange - imagine the following case:
1. A product has a URL in an input field
1. Editor thinks that this URL is bad, it shall be newly generated
1. Editor removes content if URL field
1. Editor reloads and is happy
But actually the new value is only generated in `getDataForEditmode` - in the database the value has not changed. This is quite confusing as described in #4600.

**Comment by brusch** [comment]
@BlackbitNeueMedien we have recently made some changes regarding default values, see #5594 .
I think this change is also relevant for this PR in terms of the nullify question. 
This PR should also implement the generator for all other data-types that support default values. 
Thanks a lot!

File: models/DataObject/ClassDefinition/Helper/DefaultValueGeneratorResolver.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+{
+    public static $cache = [];
+
+    /**
+     * @param $generatorClass
+     *
+     * @return DefaultValueGeneratorInterface
+     */
+    public static function resolveGenerator($generatorClass)
```
Comment (weisswurstkanone) [inline models/DataObject/ClassDefinition/Helper/DefaultValueGeneratorResolver.php]: There is already Pimcore\Model\DataObject\ClassDefinition\Helper\ClassResolver::resolve
which does the same magic

**Comment by weisswurstkanone** [comment]
Could you please update the dev documentation as well?

**Comment by weisswurstkanone** [comment]
@BlackbitNeueMedien ping. would be nice if we could add this to the next release

**Comment by brusch** [comment]
Ping 🏓
Quite old already, so I guess it would be good to get it out of the way 😊

File: models/DataObject/ClassDefinition/Helper/DefaultValueGeneratorResolver.php (hunk 2 above)
Comment (BlackbitNeueMedien) [reply models/DataObject/ClassDefinition/Helper/DefaultValueGeneratorResolver.php]: Actually this method does use `Pimcore\Model\DataObject\ClassDefinition\Helper\ClassResolver::resolve()` - what change are you thinking of?

Comment (weisswurstkanone) [reply models/DataObject/ClassDefinition/Helper/DefaultValueGeneratorResolver.php]: Comment is outdated then. Noticed that there are still changes going on. Please let me know when you are done. Thanks a lot!

File: models/DataObject/ClassDefinition/Data/Input.php [hunk 3]
```diff
@@ -38,6 +39,11 @@ class Input extends Data implements ResourcePersistenceAwareInterface, QueryReso
      */
     public $width;
 
+    /**
+     * @var string
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/Input.php]:
```suggestion
     * @var string|null
```

File: models/DataObject/ClassDefinition/Data/Input.php [hunk 4]
```diff
@@ -300,4 +308,37 @@ public function isFilterable(): bool
...
+     *
+     * @return null|int
+     */
+    protected function doGetDefaultValue($object, $context = [])
+    {
+        return $this->getDefaultValue();
+    }
+
+    /**
+     * @return int|null
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/Input.php]:
```suggestion
     * @return string|null
```

File: models/DataObject/ClassDefinition/Data/Input.php [hunk 5]
```diff
@@ -300,4 +308,37 @@ public function isFilterable(): bool
...
+     */
+    public function getDefaultValue()
+    {
+        return $this->defaultValue;
+    }
+
+    /**
+     * @param int $defaultValue
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/Input.php]:
```suggestion
     * @param string $defaultValue
```

**Comment by BlackbitNeueMedien** [comment]
@weisswurstkanone #5850 is about unique fields. This PR is about default values. I listed the data types which currently use `\Pimcore\Model\DataObject\Traits\DefaultValueTrait`.

**Comment by weisswurstkanone** [comment]
Yes, but I noticed that the line talking about unique fields was touched as well. Maybe just a re-formatting thing.

**Comment by BlackbitNeueMedien** [comment]
Default value generator would also make sense for URL slug data type, wouldn't it? I actually implemented this feature for URL slugs based on an object's current data - but as url slug datatype did not exist at that time, I used input.

**Comment by weisswurstkanone** [comment]
Makes definitely sense and would be a nice improvement. Please add it to this PR if possible 👌

**Comment by brusch** [comment]
@BlackbitNeueMedien any updates? 😊 Thanks a lot!

**Comment by BlackbitNeueMedien** [comment]
Not much time at the moment. Could you review it without the support of URL slug data type?

**Comment by weisswurstkanone** [comment]
Sure, created a follow-up ticket https://github.com/pimcore/pimcore/issues/6132
@BlackbitNeueMedien thanks so far

**Comment by weisswurstkanone** [comment]
Could you please check or explain why the $context information is completely different when compared to a dynamic options provider for example?

image

For the default value generator it looks like this (fields are on the same level directly under the object).

image

File: bundles/AdminBundle/Resources/public/js/pimcore/object/classes/data/checkbox.js [hunk 6]
```diff
@@ -82,7 +82,16 @@ pimcore.object.classes.data.checkbox = Class.create(pimcore.object.classes.data.
         });
 
         return [
-            defaultField, {
+            defaultField,
+            {
```
Comment (weisswurstkanone) [inline bundles/AdminBundle/Resources/public/js/pimcore/object/classes/data/checkbox.js:86]: new field should also be considered when cloning the checkbox (same for other data types). see method "applySpecialData"

**Comment by weisswurstkanone** [comment]
@BlackbitNeueMedien ping

File: bundles/AdminBundle/Resources/public/js/pimcore/object/classes/data/checkbox.js (hunk 6 above)
Comment (weisswurstkanone) [reply bundles/AdminBundle/Resources/public/js/pimcore/object/classes/data/checkbox.js:86]: added support for clone

**Pull request merged**